File: mod/ondb/lib/py/ondb_core.py
"""ondb core: parse TSV and provide query methods."""
import sys

class Ondb:
    def __init__(self):
        self.entities = {}   # id -> {type, ctime, mtime, alive, props, order}
        self.links = []      # [{from, rel, to, epoch, props}]
        self._entity_order = []
        self._link_map = {}  # (from, rel, to) -> link

    def load(self, lines):
        for line in lines:
            line = line.rstrip('\n')
            if not line:
                continue
            cols = line.split('\t')
            op = cols[0]
            if op == 'add':
                self._do_add(cols[1], cols[2], int(cols[3]), cols[4:])
            elif op == 'set':
                self._do_set(cols[1], int(cols[2]), cols[3:])
            elif op == 'rm':
                self._do_rm(cols[1])
            elif op == 'link':
                self._do_link(cols[1], cols[2], cols[3], int(cols[4]), cols[5:])
            elif op == 'unlink':
                self._do_unlink(cols[1], cols[2], cols[3])

    def _parse_props(self, cols):
        props = {}
        for i in range(0, len(cols), 2):
            if i + 1 < len(cols):
                k = cols[i]
                v = cols[i + 1]
                props[k] = v
        return props

    def _do_add(self, typ, eid, ts, props_cols):
        if eid in self.entities:
            return
        self.entities[eid] = {
            'type': typ, 'ctime': ts, 'mtime': ts,
            'alive': True, 'props': self._parse_props(props_cols),
            'order': len(self._entity_order)
        }
        self._entity_order.append(eid)

    def _do_set(self, eid, ts, props_cols):
        e = self.entities.get(eid)
        if not e or not e['alive']:
            return
        e['mtime'] = ts
        for i in range(0, len(props_cols), 2):
            if i + 1 < len(props_cols):
                k = props_cols[i]
                v = props_cols[i + 1]
                if v:
                    e['props'][k] = v
                else:
                    e['props'].pop(k, None)

    def _do_rm(self, eid):
        e = self.entities.get(eid)
        if e:
            e['alive'] = False

    def _do_link(self, fr, rel, to, ts, props_cols):
        key = (fr, rel, to)
        if key in self._link_map:
            return
        link = {'from': fr, 'rel': rel, 'to': to, 'epoch': ts,
                'props': self._parse_props(props_cols)}
        self.links.append(link)
        self._link_map[key] = link

    def _do_unlink(self, fr, rel, to):
        key = (fr, rel, to)
        if key in self._link_map:
            link = self._link_map.pop(key)
            link['from'] = None

    def ls(self, etype=None):
        out = []
        for eid in self._entity_order:
            e = self.entities[eid]
            if not e['alive']:
                continue
            if etype and e['type'] != etype:
                continue
            out.append((eid, e['type'], e['props'].get('name', '')))
        return out

    def linked(self, eid, rel=None, direction='outgoing'):
        out = []
        for link in self.links:
            if link['from'] is None:
                continue
            if direction == 'outgoing':
                if link['from'] != eid:
                    continue
            elif direction == 'incoming':
                if link['to'] != eid:
                    continue
            else:
                if link['from'] != eid and link['to'] != eid:
                    continue
            if rel and link['rel'] != rel:
                continue
            target = link['to'] if direction == 'outgoing' else link['from']
            e = self.entities.get(target)
            et = e['type'] if e and e['alive'] else ''
            out.append((link['rel'], target, et, link['props']))
        return out

    def get(self, eid):
        e = self.entities.get(eid)
        if not e or not e['alive']:
            return None
        return {
            'id': eid, 'type': e['type'],
            'ctime': e['ctime'], 'mtime': e['mtime'],
            'props': dict(e['props'])
        }

    def query(self, etype):
        return self.ls(etype)

    def related(self, eid, rel=None, direction='outgoing'):
        """Return related entities with full info (like ontology.py get_related)."""
        out = []
        for link in self.links:
            if link['from'] is None:
                continue
            if direction == 'outgoing':
                if link['from'] != eid:
                    continue
            elif direction == 'incoming':
                if link['to'] != eid:
                    continue
            else:
                if link['from'] != eid and link['to'] != eid:
                    continue
            if rel and link['rel'] != rel:
                continue
            target = link['to'] if direction == 'outgoing' else link['from']
            e = self.entities.get(target)
            if not e or not e['alive']:
                continue
            out.append({
                'relation': link['rel'],
                'direction': direction,
                'entity': {
                    'id': target,
                    'type': e['type'],
                    'props': dict(e['props'])
                }
            })
        return out

    def validate(self):
        """Basic validation: check dangling links."""
        errors = []
        for link in self.links:
            if link['from'] is None:
                continue
            if link['from'] not in self.entities or not self.entities[link['from']]['alive']:
                errors.append(f"dangling_from: link from={link['from']} rel={link['rel']} to={link['to']}")
            if link['to'] not in self.entities or not self.entities[link['to']]['alive']:
                errors.append(f"dangling_to: link from={link['from']} rel={link['rel']} to={link['to']}")
        return errors

File: mod/ondb/lib/js/ondb_core.js
/**
 * ondb core: parse TSV and provide query methods.
 * Optimized for bun runtime.
 */

class Ondb {
  constructor() {
    this.entities = new Map();
    this.links = [];
    this._entity_order = [];
    this._link_map = new Map();
  }

  load(text) {
    let start = 0;
    const len = text.length;
    while (start < len) {
      let end = text.indexOf('\n', start);
      if (end === -1) end = len;
      if (end > start) {
        const line = text.slice(start, end);
        const cols = line.split('\t');
        const op = cols[0];
        if (op === 'add') {
          this._do_add(cols[1], cols[2], +cols[3], cols, 4);
        } else if (op === 'set') {
          this._do_set(cols[1], +cols[2], cols, 3);
        } else if (op === 'rm') {
          this._do_rm(cols[1]);
        } else if (op === 'link') {
          this._do_link(cols[1], cols[2], cols[3], +cols[4], cols, 5);
        } else if (op === 'unlink') {
          this._do_unlink(cols[1], cols[2], cols[3]);
        }
      }
      start = end + 1;
    }
  }

  _parse_props(cols, from) {
    const props = {};
    for (let i = from; i < cols.length; i += 2) {
      if (i + 1 < cols.length) {
        props[cols[i]] = cols[i + 1];
      }
    }
    return props;
  }

  _do_add(typ, eid, ts, cols, from) {
    if (this.entities.has(eid)) return;
    this.entities.set(eid, {
      type: typ, ctime: ts, mtime: ts,
      alive: true, props: this._parse_props(cols, from)
    });
    this._entity_order.push(eid);
  }

  _do_set(eid, ts, cols, from) {
    const e = this.entities.get(eid);
    if (!e || !e.alive) return;
    e.mtime = ts;
    for (let i = from; i < cols.length; i += 2) {
      if (i + 1 < cols.length) {
        const k = cols[i];
        const v = cols[i + 1];
        if (v) {
          e.props[k] = v;
        } else {
          delete e.props[k];
        }
      }
    }
  }

  _do_rm(eid) {
    const e = this.entities.get(eid);
    if (e) e.alive = false;
  }

  _do_link(fr, rel, to, ts, cols, from) {
    const key = fr + '\t' + rel + '\t' + to;
    if (this._link_map.has(key)) return;
    const link = { from: fr, rel, to, epoch: ts, props: this._parse_props(cols, from) };
    this.links.push(link);
    this._link_map.set(key, link);
  }

  _do_unlink(fr, rel, to) {
    const key = fr + '\t' + rel + '\t' + to;
    const link = this._link_map.get(key);
    if (link) {
      link.from = null;
      this._link_map.delete(key);
    }
  }

  ls(etype = null) {
    const out = [];
    for (let i = 0; i < this._entity_order.length; i++) {
      const eid = this._entity_order[i];
      const e = this.entities.get(eid);
      if (!e || !e.alive) continue;
      if (etype && e.type !== etype) continue;
      out.push([eid, e.type, e.props.name || '']);
    }
    return out;
  }

  get(eid) {
    const e = this.entities.get(eid);
    if (!e || !e.alive) return null;
    return {
      id: eid, type: e.type,
      ctime: e.ctime, mtime: e.mtime,
      props: { ...e.props }
    };
  }

  query(etype = null) {
    return this.ls(etype);
  }

  related(eid, rel = null, direction = 'outgoing') {
    const out = [];
    for (let i = 0; i < this.links.length; i++) {
      const link = this.links[i];
      if (link.from === null) continue;
      if (direction === 'outgoing') {
        if (link.from !== eid) continue;
      } else if (direction === 'incoming') {
        if (link.to !== eid) continue;
      } else {
        if (link.from !== eid && link.to !== eid) continue;
      }
      if (rel && link.rel !== rel) continue;
      const target = direction === 'outgoing' ? link.to : link.from;
      const e = this.entities.get(target);
      if (!e || !e.alive) continue;
      out.push({
        relation: link.rel,
        direction,
        entity: { id: target, type: e.type, props: { ...e.props } }
      });
    }
    return out;
  }

  validate() {
    const errors = [];
    for (let i = 0; i < this.links.length; i++) {
      const link = this.links[i];
      if (link.from === null) continue;
      const from_e = this.entities.get(link.from);
      const to_e = this.entities.get(link.to);
      if (!from_e || !from_e.alive) {
        errors.push(`dangling_from: link from=${link.from} rel=${link.rel} to=${link.to}`);
      }
      if (!to_e || !to_e.alive) {
        errors.push(`dangling_to: link from=${link.from} rel=${link.rel} to=${link.to}`);
      }
    }
    return errors;
  }
}

module.exports = { Ondb };
